File: repository-hbase/src/main/java/hbase/repository/HBaseAdminDataRepository.java
package hbase.repository;

import hbase.connector.HBaseConnector;
import hbase.util.RowKeyUtils;
import model.AdminData;
import net.jodah.failsafe.CircuitBreaker;
import net.jodah.failsafe.Failsafe;
import org.apache.hadoop.hbase.Cell;
import org.apache.hadoop.hbase.CellUtil;
import org.apache.hadoop.hbase.TableName;
import org.apache.hadoop.hbase.client.Get;
import org.apache.hadoop.hbase.client.Result;
import org.apache.hadoop.hbase.client.Table;
import org.apache.hadoop.hbase.util.Bytes;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import repository.AdminDataRepository;

import javax.inject.Inject;
import java.time.Month;
import java.time.YearMonth;
import java.util.Optional;
import java.util.concurrent.CompletionStage;
import java.util.concurrent.Executor;
import java.util.concurrent.TimeUnit;

import static java.util.concurrent.CompletableFuture.supplyAsync;

/**
 * HBase implementation of an Admin Data DAO
 */
public class HBaseAdminDataRepository implements AdminDataRepository {

    private static final Logger LOG = LoggerFactory.getLogger(HBaseAdminDataRepository.class.getName());
    private static final YearMonth HARDCODED_CURRENT_PERIOD = YearMonth.of(2017, Month.JUNE);
    private final CircuitBreaker circuitBreaker = new CircuitBreaker().withDelay(1, TimeUnit.MINUTES).withFailureThreshold(1).withSuccessThreshold(3);

    private final Executor ec;
    private final HBaseConnector connector;
    private final TableName tableName;

    @Inject
    public HBaseAdminDataRepository(HBaseConnector connector, Executor ec) {
        this.connector = connector;
        this.ec = ec;
        this.tableName = TableName.valueOf(System.getProperty("hbase.namespace", ""), System.getProperty("hbase.table", "data"));
    }

    @Override
    public CompletionStage<YearMonth> getCurrentPeriod() {
        //TODO: Implement retrieving the current period from HBase
        return supplyAsync(() -> Failsafe.with(circuitBreaker).get(() -> HARDCODED_CURRENT_PERIOD), ec);
    }

    @Override
    public CompletionStage<Optional<AdminData>> lookup(YearMonth referencePeriod, String key) {
        return supplyAsync(() -> Failsafe.with(circuitBreaker).get(() -> getAdminData(referencePeriod, key)), ec);
    }

    private Optional<AdminData> getAdminData(YearMonth referencePeriod, String key) throws Exception {
        String rowKey = RowKeyUtils.createRowKey(referencePeriod, key);
        Optional<AdminData> adminData;
        try (Table table = connector.getConnection().getTable(tableName)) {
            Get get = new Get(Bytes.toBytes(rowKey));
            Result result = table.get(get);
            if (result.isEmpty()) {
                LOG.debug("No data found for row key '{}'", rowKey);
                adminData = Optional.empty();
            } else {
                LOG.debug("Found data for row key '{}'", rowKey);
                adminData = Optional.of(convertToAdminData(result));
            }
        } catch (Exception e) {
            LOG.error("Error getting data for row key '{}'", rowKey, e);
            throw e;
        }
        return adminData;
    }

    private AdminData convertToAdminData(Result result) {
        AdminData adminData = RowKeyUtils.createAdminDataFromRowKey(Bytes.toString(result.getRow()));
        for (Cell cell : result.listCells()) {
            String column = new String(CellUtil.cloneQualifier(cell));
            String value = new String(CellUtil.cloneValue(cell));
            LOG.debug("Found data column '{}' with value '{}'", column, value);
            adminData.putVariable(column, value);
        }
        return adminData;
    }

}

File: repository-hbase/src/main/scala/hbase/repository/HBaseAdminDataRepository.scala
package hbase.repository

import javax.inject.Inject

import scala.collection.JavaConversions._
import scala.concurrent.{ ExecutionContext, Future }
import scala.util.{ Failure, Success, Try }

import play.api.Configuration
import play.api.http.{ ContentTypes, Status }
import play.api.libs.json.{ JsValue, Json }
import play.api.mvc.{ Results, Result => PlayResult }
import org.apache.hadoop.hbase.CellUtil
import org.apache.hadoop.hbase.client.{ Result, _ }
import org.apache.hadoop.hbase.util.Bytes
import org.slf4j.{ Logger, LoggerFactory }
import com.github.nscala_time.time.Imports.{ DateTimeFormat, YearMonth }
import com.netaporter.uri.dsl._

import hbase.connector.HBaseConnector
import hbase.model.AdminData
import hbase.util.{ HBaseConfig, RowKeyUtils }

import services.util.EncodingUtil.{ decodeBase64, encodeBase64 }
import services.websocket.RequestGenerator

/**
 * HBaseAdminDataRepository
 * ----------------
 * Author: haqa
 * Date: 20 November 2017 - 11:58
 * Copyright (c) 2017  Office for National Statistics
 */
class HBaseAdminDataRepository @Inject() (
  val connector: HBaseConnector,
  ws: RequestGenerator,
  val configuration: Configuration) extends AdminDataRepository with Status with Results with ContentTypes with HBaseConfig {

  implicit val ec = ExecutionContext.global

  private final val logger: Logger = LoggerFactory.getLogger(getClass.getName)
  private final val OPEN_ALERT = "----- circuit breaker opened! -----"
  private final val CLOSED_ALERT = "----- circuit breaker closed! -----"
  private final val HALF_OPEN_ALERT = "----- circuit breaker half-open -----"

  private val maxResultSize: Long = 12L
  private val auth = encodeBase64(Seq(username, password))

  // @TODO - fix + add circuit-breaker
  @deprecated("Migrated to getAdminData with three param", "13 Dec 2017 - feature/HBase-Rest")
  override def lookup(referencePeriod: Option[YearMonth], key: String): Future[Option[AdminData]] =
    Future.successful(getAdminData(referencePeriod, key))

  override def lookup(referencePeriod: Option[YearMonth], key: String, max: Long = maxResultSize): Future[Option[Seq[AdminData]]] =
    Future.successful(getAdminData(referencePeriod, key, max))

  @deprecated("Migrated to getAdminData with three param", "13 Dec 2017 - feature/HBase-Rest")
  override def lookup(key: String, referencePeriod: Option[YearMonth]): Future[PlayResult] =
    getAdminData(key, referencePeriod)

  @deprecated("Migrated to getAdminData with three param", "13 Dec 2017 - feature/HBase-Rest")
  @throws(classOf[Exception])
  private def getAdminData(referencePeriod: Option[YearMonth], key: String): Option[AdminData] = {
    val rowKey = RowKeyUtils.createRowKey(referencePeriod.getOrElse(YearMonth.parse(
      "201706",
      DateTimeFormat.forPattern(AdminData.REFERENCE_PERIOD_FORMAT))), key)
    Try(connector.getConnection.getTable(tableName)) match {
      case Success(table: Table) =>
        table.get(new Get(Bytes.toBytes(rowKey))) match {
          case res if res.isEmpty =>
            logger.debug("No data found for row key '{}'", rowKey)
            None
          case result =>
            logger.debug("Found data for row key '{}'", rowKey)
            Some(convertToAdminData(result))
        }
      case Failure(e: Throwable) =>
        logger.error(s"Error getting data for row key $rowKey", e)
        throw e
    }
  }

  @throws(classOf[Exception])
  private def getAdminData(referencePeriod: Option[YearMonth], key: String, max: Long): Option[Seq[AdminData]] = {
    Try(connector.getConnection.getTable(tableName)) match {
      case Success(table: Table) =>
        referencePeriod match {
          case Some(y: YearMonth) =>
            val rowKey = RowKeyUtils.createRowKey(referencePeriod.getOrElse(y), key)
            table.get(new Get(Bytes.toBytes(rowKey))) match {
              case res if res.isEmpty =>
                logger.debug("No data found for row key '{}'", rowKey)
                None
              case result =>
                logger.debug("Found data for row key '{}'", rowKey)
                Some(Seq(convertToAdminData(result)))
            }
          case None =>
            val scan = new Scan()
              .setReversed(true)
              .setRowPrefixFilter(Bytes.toBytes(key))
            scan.setMaxResultSize(2)
            Option(table.getScanner(scan).next) match {
              case Some(_) => Some(table.getScanner(scan).map { x => convertToAdminData(x) }.toSeq)
              case None => None
            }
        }
      case Failure(e: Exception) =>
        logger.error(s"Error getting data for row key $key", e)
        throw e
    }
  }

  @throws(classOf[Exception])
  private def getAdminData(key: String, referencePeriod: Option[YearMonth], max: Long = maxResultSize): Future[PlayResult] = {
    val headers = Seq("Accept" -> "application/json", "Authorization" -> s"Basic $auth")
    val request = referencePeriod match {
      case Some(r: YearMonth) =>
        val rowKey = RowKeyUtils.createRowKey(r, key)
        val uri = baseUrl / tableName.getNameWithNamespaceInclAsString / rowKey / columnFamily
        ws.singleGETRequest(uri.toString, headers)
      case None =>
        val params = Seq("reversed" -> "true", "limit" -> max.toString)
        val uri = baseUrl / tableName.getNameWithNamespaceInclAsString / key + "*"
        ws.singleGETRequest(uri.toString, headers, params)
    }
    request.map {
      case response if response.status == OK => {
        val resp = (response.json \ "Row").as[Seq[JsValue]]
        Try(resp.map(v => convertToAdminData(v))) match {
          case Success(adminData: Seq[AdminData]) =>
            //            Some(adminData)
            Ok(Json.toJson(adminData)).as(JSON)
          case Failure(ex) =>
            // TODO - add exception type
            //            None
            BadRequest(s"$ex")
        }
      }
      case response if response.status == NOT_FOUND => NotFound(response.body).as(JSON)
    }
  }

  private def convertToAdminData(result: JsValue): AdminData = {
    val key = (result \ "key").as[String]
    logger.debug(s"Found record $key")
    val adminData: AdminData = RowKeyUtils.createAdminDataFromRowKey(decodeBase64(key))
    val varMap = (result \ "Cell").as[Seq[JsValue]].map { cell =>
      val column = decodeBase64((cell \ "column").as[String])
      val value = decodeBase64((cell \ "$").as[String])
      column -> value
    }.toMap
    val newPutAdminData = adminData.putVariable(varMap)
    newPutAdminData
  }

  private def convertToAdminData(result: Result): AdminData = {
    val adminData: AdminData = RowKeyUtils.createAdminDataFromRowKey(Bytes.toString(result.getRow))
    val varMap = result.listCells.toList.map { cell =>
      val column = new String(CellUtil.cloneQualifier(cell))
      val value = new String(CellUtil.cloneValue(cell))
      column -> value
    }.toMap
    val newPutAdminData = adminData.putVariable(varMap)
    newPutAdminData
  }

}

